rdr_service/storage.py

```diff
@@ -370,3 +372,102 @@ def get_storage_provider():
     provider_class = StorageProvider.get_provider(default=default_provider)
     return provider_class()
 
+
+class GoogleCloudStorageCSVReader:
+    """
+    Read a CSV file from a bucket and yield rows or chunks of rows.
+    """
+    def __init__(self, cloud_csv_file, column_names:list=None, delimiter=','):
+        """
+        :param cloud_csv_file: bucket and file blob name of cloud csv file.
+        :param column_names: list of column names if first row does not have the col names.
+        """
+        if not cloud_csv_file or not isinstance(cloud_csv_file, str):
+            raise ValueError('Invalid cloud storage csv file name.')
+
+        self.cloud_csv_file = cloud_csv_file
+        self.column_names = column_names
+        self._delimiter = delimiter
+        self._tmp_file = None  # Temp file object handle.
+        self._strio = None  # StringIO object
+        self._header = None  # delimited string of column names.
+        self._reader = None  # csv.DictReader object.
+        self._batch_size = 1000  # number of rows from temp file to read.
```

Comment:
small concern: It still looks like a DictReader goes line by line on it's own without our intervention. This seems like it would increase memory compared to that, since we're reading 1000 lines into memory at once.

Reply:
We are never "comprehending" the entire 1000 records from the DictReader at one time in the class.  If the DictReader does what you said the other day, it should only be serving up one record at a time in this code.   But in a worse case scenario, the DictReader can only load 1,000 records into memory at any given time inside this class.

Reply:
The DictReady would function in the way that it's been implemented and optimized. But I was specifically referring to this class. This implementation would hit the worst case on its own in the `_read_batch_from_file` by reading 1000 lines into memory at once.

Reply:
In the StringIO object yes, but that is a byte array.  So lets say the average row is 50 bytes long in the csv file. The memory the StringIO would use can be calculated at 50 * 1,000.  So 50k + object overhead, that's not a lot of memory used there. If that is the worst case, I would take that any day.  What I am trying to work towards here is not to use the bare minimum memory, but rather to use a roughly fixed amount of memory used that should never cause memory pressure issues like we have seen.